<a href="https://colab.research.google.com/github/Daneswariaddala/Daneswari/blob/main/E_Commerce_orders.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
orders = pd.read_csv('/content/List of Orders.csv')
details = pd.read_csv('/content/Order Details.csv')
target = pd.read_csv('/content/Sales target.csv')

In [2]:
orders

Order ID  Order Date CustomerName           State       City
0    B-25601  01-04-2018       Bharat         Gujarat  Ahmedabad
1    B-25602  01-04-2018        Pearl     Maharashtra       Pune
2    B-25603  03-04-2018        Jahan  Madhya Pradesh     Bhopal
3    B-25604  03-04-2018       Divsha       Rajasthan     Jaipur
4    B-25605  05-04-2018      Kasheen     West Bengal    Kolkata
..       ...         ...          ...             ...        ...
555      NaN         NaN          NaN             NaN        NaN
556      NaN         NaN          NaN             NaN        NaN
557      NaN         NaN          NaN             NaN        NaN
558      NaN         NaN          NaN             NaN        NaN
559      NaN         NaN          NaN             NaN        NaN

[560 rows x 5 columns]

In [3]:
details

Order ID  Amount  Profit  Quantity     Category      Sub-Category
0     B-25601  1275.0 -1148.0         7    Furniture         Bookcases
1     B-25601    66.0   -12.0         5     Clothing             Stole
2     B-25601     8.0    -2.0         3     Clothing       Hankerchief
3     B-25601    80.0   -56.0         4  Electronics  Electronic Games
4     B-25602   168.0  -111.0         2  Electronics            Phones
...       ...     ...     ...       ...          ...               ...
1495  B-26099   835.0   267.0         5  Electronics            Phones
1496  B-26099  2366.0   552.0         5     Clothing          Trousers
1497  B-26100   828.0   230.0         2    Furniture            Chairs
1498  B-26100    34.0    10.0         2     Clothing           T-shirt
1499  B-26100    72.0    16.0         2     Clothing             Shirt

[1500 rows x 6 columns]

In [4]:
target

Month of Order Date     Category   Target
0               Apr-18    Furniture  10400.0
1               May-18    Furniture  10500.0
2               Jun-18    Furniture  10600.0
3               Jul-18    Furniture  10800.0
4               Aug-18    Furniture  10900.0
5               Sep-18    Furniture  11000.0
6               Oct-18    Furniture  11100.0
7               Nov-18    Furniture  11300.0
8               Dec-18    Furniture  11400.0
9               Jan-19    Furniture  11500.0
10              Feb-19    Furniture  11600.0
11              Mar-19    Furniture  11800.0
12              Apr-18     Clothing  12000.0
13              May-18     Clothing  12000.0
14              Jun-18     Clothing  12000.0
15              Jul-18     Clothing  14000.0
16              Aug-18     Clothing  14000.0
17              Sep-18     Clothing  14000.0
18              Oct-18     Clothing  16000.0
19              Nov-18     Clothing  16000.0
20              Dec-18     Clothing  16000.0
21              Jan-19     Clothing  16000.0
22              Feb-19     Clothing  16000.0
23              Mar-19     Clothing  16000.0
24              Apr-18  Electronics   9000.0
25              May-18  Electronics   9000.0
26              Jun-18  Electronics   9000.0
27              Jul-18  Electronics   9000.0
28              Aug-18  Electronics   9000.0
29              Sep-18  Electronics   9000.0
30              Oct-18  Electronics   9000.0
31              Nov-18  Electronics   9000.0
32              Dec-18  Electronics   9000.0
33              Jan-19  Electronics  16000.0
34              Feb-19  Electronics  16000.0
35              Mar-19  Electronics  16000.0

**List of orders table**:  
List of orders contains 560 observations acroess 5 variables

**Order.ID** is chr type; it is combined from letter and order number and should be unique for each row; contains 500 unique values and 60 missing values  
**OrderDate** is chr type, but we would like to transform it into the date data type; contains 12 missing values  
**CustomerName** is chr type; contains 10 missing values  
**State** is chr type; contains 12 missing values  
**City** is chr type; contains 8 missing values  
We do not have a country column, but we know that all records come from India, so we can add country to the table to keep our data cohesive

In [5]:
# List of orders
print(str(orders)) # Quick preview of a table
print("Unique orders: ", len(orders['Order ID'].unique())) # Count distinct values in column 'Order ID'
print("Missing values:", sum(orders['Order ID'].isna())) # Sum of missing values in column 'Order ID'

    Order ID  Order Date CustomerName           State       City
0    B-25601  01-04-2018       Bharat         Gujarat  Ahmedabad
1    B-25602  01-04-2018        Pearl     Maharashtra       Pune
2    B-25603  03-04-2018        Jahan  Madhya Pradesh     Bhopal
3    B-25604  03-04-2018       Divsha       Rajasthan     Jaipur
4    B-25605  05-04-2018      Kasheen     West Bengal    Kolkata
..       ...         ...          ...             ...        ...
555      NaN         NaN          NaN             NaN        NaN
556      NaN         NaN          NaN             NaN        NaN
557      NaN         NaN          NaN             NaN        NaN
558      NaN         NaN          NaN             NaN        NaN
559      NaN         NaN          NaN             NaN        NaN

[560 rows x 5 columns]
Unique orders:  501
Missing values: 60


**Order details table** :  
Order details table contains 1500 observations across 6 variables

**Order.ID** is char type, format is the same as in the List of orders table; contains 500 unique values, no missing values  
**Amount** is a numerical type, it should be > 0; all values are above 0, no missing values  
**Profit** is a numerical type, it can be below or above 0; no missing values  
**Quantity** integer data type, should be above 0; no missing values, all observations are > 0, as demanded  
**Category** is a chr type, it shoud contain 3 main categories: "Furniture", "Clothing", "Electronics"; no missing values, all category types are correct  
**Sub.Category** contains 17 categories; no missing values, all categories names are as stated in dataset description

In [6]:
# order details
print(str(details))  # Quick preview of a table
print("Unique orders: ", len(details['Order ID'].unique()))  # Count distinct values in column 'Order ID'
print("Missing values:", sum(details['Order ID'].isna()))  # Sum of missing values in column 'Order ID'

     Order ID  Amount  Profit  Quantity     Category      Sub-Category
0     B-25601  1275.0 -1148.0         7    Furniture         Bookcases
1     B-25601    66.0   -12.0         5     Clothing             Stole
2     B-25601     8.0    -2.0         3     Clothing       Hankerchief
3     B-25601    80.0   -56.0         4  Electronics  Electronic Games
4     B-25602   168.0  -111.0         2  Electronics            Phones
...       ...     ...     ...       ...          ...               ...
1495  B-26099   835.0   267.0         5  Electronics            Phones
1496  B-26099  2366.0   552.0         5     Clothing          Trousers
1497  B-26100   828.0   230.0         2    Furniture            Chairs
1498  B-26100    34.0    10.0         2     Clothing           T-shirt
1499  B-26100    72.0    16.0         2     Clothing             Shirt

[1500 rows x 6 columns]
Unique orders:  500
Missing values: 0


***Sales target table : ***  
Sales target table contains of 36 observations of 3 variables

**Month.of.Order.Date** is chr data type, contains month and year parts for period since April 2018 till March 2019 for 3 categories(12 months * 3 categories = 36 observations), it should be transformed into date type for further processing; no missing values, all values fit dataset description  
**Category** chr data rype, contains 3 main categories; no missing values, all categories are correct  
**Target** num type; no missing values, all values are > 0

In [7]:
# sales target
print(str(target))  # Quick preview of a table

# Check the actual column names in the DataFrame
print(target.columns)

# Replace 'Month.of.Order.Date' with the correct column name
# For example, if the actual column name is 'Month of Order Date', use:
print("Unique months/categories: ", len(target['Month of Order Date'].unique()))

# If the column name is different, adjust accordingly.
print("Missing values in target table:", target.isnull().sum().sum()) # Total missing values in the DataFrame

   Month of Order Date     Category   Target
0               Apr-18    Furniture  10400.0
1               May-18    Furniture  10500.0
2               Jun-18    Furniture  10600.0
3               Jul-18    Furniture  10800.0
4               Aug-18    Furniture  10900.0
5               Sep-18    Furniture  11000.0
6               Oct-18    Furniture  11100.0
7               Nov-18    Furniture  11300.0
8               Dec-18    Furniture  11400.0
9               Jan-19    Furniture  11500.0
10              Feb-19    Furniture  11600.0
11              Mar-19    Furniture  11800.0
12              Apr-18     Clothing  12000.0
13              May-18     Clothing  12000.0
14              Jun-18     Clothing  12000.0
15              Jul-18     Clothing  14000.0
16              Aug-18     Clothing  14000.0
17              Sep-18     Clothing  14000.0
18              Oct-18     Clothing  16000.0
19              Nov-18     Clothing  16000.0
20              Dec-18     Clothing  16000.0
21        

**Data cleaning and transformation**

In [ ]:
# Add a 'Country' column with the value 'India' to the 'orders' DataFrame
orders['Country'] = 'India'

# Remove rows with null values in any column of the 'orders' DataFrame
orders = orders.dropna()

orders = orders.rename(columns={"Order.ID": "order_id"})
orders

Order ID  Order Date CustomerName              State       City Country
0    B-25601  01-04-2018       Bharat            Gujarat  Ahmedabad   India
1    B-25602  01-04-2018        Pearl        Maharashtra       Pune   India
2    B-25603  03-04-2018        Jahan     Madhya Pradesh     Bhopal   India
3    B-25604  03-04-2018       Divsha          Rajasthan     Jaipur   India
4    B-25605  05-04-2018      Kasheen        West Bengal    Kolkata   India
..       ...         ...          ...                ...        ...     ...
495  B-26096  28-03-2019       Atharv        West Bengal    Kolkata   India
496  B-26097  28-03-2019         Vini          Karnataka  Bangalore   India
497  B-26098  29-03-2019        Pinky  Jammu and Kashmir    Kashmir   India
498  B-26099  30-03-2019       Bhishm        Maharashtra     Mumbai   India
499  B-26100  31-03-2019       Hitika     Madhya Pradesh     Indore   India

[500 rows x 6 columns]

In [ ]:
details = details.rename(
    columns={
        "Order.ID": "order_id",
        "Amount": "amount",
        "Profit": "profit",
        "Quantity": "quantity",
        "Category": "category",
        "Sub.Category": "sub_category"
    }
)

details

Order ID  amount  profit  quantity     category      Sub-Category
0     B-25601  1275.0 -1148.0         7    Furniture         Bookcases
1     B-25601    66.0   -12.0         5     Clothing             Stole
2     B-25601     8.0    -2.0         3     Clothing       Hankerchief
3     B-25601    80.0   -56.0         4  Electronics  Electronic Games
4     B-25602   168.0  -111.0         2  Electronics            Phones
...       ...     ...     ...       ...          ...               ...
1495  B-26099   835.0   267.0         5  Electronics            Phones
1496  B-26099  2366.0   552.0         5     Clothing          Trousers
1497  B-26100   828.0   230.0         2    Furniture            Chairs
1498  B-26100    34.0    10.0         2     Clothing           T-shirt
1499  B-26100    72.0    16.0         2     Clothing             Shirt

[1500 rows x 6 columns]

In [9]:
merge_orders = pd.read_csv('/content/List of Orders.csv')


In [10]:
merge_details = pd.read_csv('/content/Order Details.csv')

In [11]:
merge_target = pd.read_csv('/content/Sales target.csv')

In [12]:
merge_all = merge_orders.merge(merge_details,how='outer')

In [13]:
merge_all = merge_all.merge(merge_target, how='outer')

In [14]:
merge_all.to_csv("merge_all_dataframes.csv", index=False)

In [16]:
data = pd.read_csv('/content/merge_all_dataframes.csv')

In [17]:
data

Order ID  Order Date CustomerName    State       City  Amount  Profit  \
0      B-25601  01-04-2018       Bharat  Gujarat  Ahmedabad    66.0   -12.0   
1      B-25601  01-04-2018       Bharat  Gujarat  Ahmedabad    66.0   -12.0   
2      B-25601  01-04-2018       Bharat  Gujarat  Ahmedabad    66.0   -12.0   
3      B-25601  01-04-2018       Bharat  Gujarat  Ahmedabad    66.0   -12.0   
4      B-25601  01-04-2018       Bharat  Gujarat  Ahmedabad    66.0   -12.0   
...        ...         ...          ...      ...        ...     ...     ...   
18055      NaN         NaN          NaN      NaN        NaN     NaN     NaN   
18056      NaN         NaN          NaN      NaN        NaN     NaN     NaN   
18057      NaN         NaN          NaN      NaN        NaN     NaN     NaN   
18058      NaN         NaN          NaN      NaN        NaN     NaN     NaN   
18059      NaN         NaN          NaN      NaN        NaN     NaN     NaN   

       Quantity  Category Sub-Category Month of Order Date   Target  
0           5.0  Clothing        Stole              Apr-18  12000.0  
1           5.0  Clothing        Stole              May-18  12000.0  
2           5.0  Clothing        Stole              Jun-18  12000.0  
3           5.0  Clothing        Stole              Jul-18  14000.0  
4           5.0  Clothing        Stole              Aug-18  14000.0  
...         ...       ...          ...                 ...      ...  
18055       NaN       NaN          NaN                 NaN      NaN  
18056       NaN       NaN          NaN                 NaN      NaN  
18057       NaN       NaN          NaN                 NaN      NaN  
18058       NaN       NaN          NaN                 NaN      NaN  
18059       NaN       NaN          NaN                 NaN      NaN  

[18060 rows x 12 columns]

In [19]:
data = data.drop_duplicates()
data

Order ID  Order Date CustomerName           State       City  Amount  \
0      B-25601  01-04-2018       Bharat         Gujarat  Ahmedabad    66.0   
1      B-25601  01-04-2018       Bharat         Gujarat  Ahmedabad    66.0   
2      B-25601  01-04-2018       Bharat         Gujarat  Ahmedabad    66.0   
3      B-25601  01-04-2018       Bharat         Gujarat  Ahmedabad    66.0   
4      B-25601  01-04-2018       Bharat         Gujarat  Ahmedabad    66.0   
...        ...         ...          ...             ...        ...     ...   
17996  B-26100  31-03-2019       Hitika  Madhya Pradesh     Indore   828.0   
17997  B-26100  31-03-2019       Hitika  Madhya Pradesh     Indore   828.0   
17998  B-26100  31-03-2019       Hitika  Madhya Pradesh     Indore   828.0   
17999  B-26100  31-03-2019       Hitika  Madhya Pradesh     Indore   828.0   
18000      NaN         NaN          NaN             NaN        NaN     NaN   

       Profit  Quantity   Category Sub-Category Month of Order Date   Target  
0       -12.0       5.0   Clothing        Stole              Apr-18  12000.0  
1       -12.0       5.0   Clothing        Stole              May-18  12000.0  
2       -12.0       5.0   Clothing        Stole              Jun-18  12000.0  
3       -12.0       5.0   Clothing        Stole              Jul-18  14000.0  
4       -12.0       5.0   Clothing        Stole              Aug-18  14000.0  
...       ...       ...        ...          ...                 ...      ...  
17996   230.0       2.0  Furniture       Chairs              Dec-18  11400.0  
17997   230.0       2.0  Furniture       Chairs              Jan-19  11500.0  
17998   230.0       2.0  Furniture       Chairs              Feb-19  11600.0  
17999   230.0       2.0  Furniture       Chairs              Mar-19  11800.0  
18000     NaN       NaN        NaN          NaN                 NaN      NaN  

[18001 rows x 12 columns]

In [21]:
data = data.dropna()
data

Order ID  Order Date CustomerName           State       City  Amount  \
0      B-25601  01-04-2018       Bharat         Gujarat  Ahmedabad    66.0   
1      B-25601  01-04-2018       Bharat         Gujarat  Ahmedabad    66.0   
2      B-25601  01-04-2018       Bharat         Gujarat  Ahmedabad    66.0   
3      B-25601  01-04-2018       Bharat         Gujarat  Ahmedabad    66.0   
4      B-25601  01-04-2018       Bharat         Gujarat  Ahmedabad    66.0   
...        ...         ...          ...             ...        ...     ...   
17995  B-26100  31-03-2019       Hitika  Madhya Pradesh     Indore   828.0   
17996  B-26100  31-03-2019       Hitika  Madhya Pradesh     Indore   828.0   
17997  B-26100  31-03-2019       Hitika  Madhya Pradesh     Indore   828.0   
17998  B-26100  31-03-2019       Hitika  Madhya Pradesh     Indore   828.0   
17999  B-26100  31-03-2019       Hitika  Madhya Pradesh     Indore   828.0   

       Profit  Quantity   Category Sub-Category Month of Order Date   Target  
0       -12.0       5.0   Clothing        Stole              Apr-18  12000.0  
1       -12.0       5.0   Clothing        Stole              May-18  12000.0  
2       -12.0       5.0   Clothing        Stole              Jun-18  12000.0  
3       -12.0       5.0   Clothing        Stole              Jul-18  14000.0  
4       -12.0       5.0   Clothing        Stole              Aug-18  14000.0  
...       ...       ...        ...          ...                 ...      ...  
17995   230.0       2.0  Furniture       Chairs              Nov-18  11300.0  
17996   230.0       2.0  Furniture       Chairs              Dec-18  11400.0  
17997   230.0       2.0  Furniture       Chairs              Jan-19  11500.0  
17998   230.0       2.0  Furniture       Chairs              Feb-19  11600.0  
17999   230.0       2.0  Furniture       Chairs              Mar-19  11800.0  

[18000 rows x 12 columns]

In [22]:
data.columns

Index(['Order ID', 'Order Date', 'CustomerName', 'State', 'City', 'Amount',
       'Profit', 'Quantity', 'Category', 'Sub-Category', 'Month of Order Date',
       'Target'],
      dtype='object')